Add explainer for MediaStream-ImageCapture spec. (#241)
w3c/mediacapture-image: explainer.md

). For manipulating video related capabilities, current settings and constraints are done via [MediaStreamTrack.applyConstraints()](https://developer.mozilla.org/en-US/docs/Web/API/MediaStreamTrack/applyConstraints). Photo-specific capabilities and current settings can be retrieved via [getPhotoCapabilities()](https://w3c.github.io/mediacapture-image/#dom-imagecapture-getphotocapabilities)/[getPhotoSettings()](https://w3c.github.io/mediacapture-image/#dom-imagecapture-getphotosettings) and configured via takePhoto()'s PhotoSettings argument. The produced image can be in the form of a [Blob](https://www.w3.org/TR/FileAPI/#dfn-Blob) or as a [ImageBitmap](https://html.spec.whatwg.org/multipage/imagebitmap-and-animations.html#imagebitmap).

## Goals

> The [MediaStream Image Capture API](https://w3c.github.io/mediacapture-image/) has been [implemented in Chrome](https://caniuse.com/#search=imageCapture) since 2017. The specification is an extension of the [Media Capture and Streams](https://w3c.github.io/mediacapture-main/) where we specify advanced features related to camera and photography as a set of property set of capabilities, constraints and settings. Most camera applications are written in native technologies to delight users with advanced use cases like controlling `exposureTime`, `focalDistance`, etc. This specification tries to bridge the gap between what is allowed on a native platform and what is exposed to the Web Platform with respect to camera capabilities. Even more advanced cameras have Pan, Tilt and Zoom support. We support those capabilities also keeping in mind the privacy and security nuances those entail and delight users of applications like video conferencing with use cases not possible before on the Web Platform.

Reviewer: ```suggestion
The [MediaStream Image Capture API](https://w3c.github.io/mediacapture-image/) has been [implemented in Chrome](https://caniuse.com/#search=imageCapture) since 2017. The specification is an extension of the [Media Capture and Streams](https://w3c.github.io/mediacapture-main/) where we specify advanced features related to camera and photography as a set of constrainable properties, with capabilities, constraints and settings. Most camera applications are written in native technologies to delight users with advanced use cases like controlling `exposureTime`, `focalDistance`, etc. This specification tries to bridge the gap between what is allowed on a native platform and what is exposed to the Web Platform with respect to camera capabilities. Even more advanced cameras have Pan, Tilt and Zoom support. We support those capabilities also keeping in mind the privacy and security nuances those entail and delight users of applications like video conferencing with use cases not possible before on the Web Platform.
```

Reviewer: Yikes github. This is basically s/set of property set of/set of constrainable properties, with/

## Methods

[takePhoto()](https://w3c.github.io/mediacapture-image/#dom-imagecapture-takephoto) returns a captured image encoded in the form of a [Blob](https://www.w3.org/TR/FileAPI/#dfn-Blob), whereas [grabFrame()](https://w3c.github.io/mediacapture-image/#dom-imagecapture-grabframe) returns a snapshot of the track video feed in the form of a non-encoded [ImageBitmap](https://html.spec.whatwg.org/multipage/imagebitmap-and-animations.html#imagebitmap). takePhoto() produces the result of a single photographic exposure using the video capture device sourcing the track and including any PhotoSettings configured, returning an encoded image in the form of a Blob if successful. grabFrame() takes a snapshot of the live video being held in track, returning an ImageBitmap if successful. Ideally takePhoto could produce the highest quality image the platform can allow for a still photograph.

## API example

```js

> var imageCapture;

Reviewer: ```suggestion
let imageCapture;
```

>   navigator.mediaDevices.getUserMedia({video: true})
>     .then(gotMedia)
>     .catch(err => console.error('getUserMedia() failed: ', err));
> 
>   function gotMedia(mediastream) {

Reviewer: I think we should use `async`/`await` in examples.
```suggestion
async function getMedia() {
  try {
    const mediastream = await navigator.mediaDevices.getUserMedia({video: true});
```

    const video = document.querySelector('video');
    video.srcObject = mediastream;

    const track = mediastream.getVideoTracks()[0];
    imageCapture = new ImageCapture(track);

    const capabilities = track.getCapabilities()
    // Check whether focus distance is supported or not.
    if (!capabilities.focusDistance) {
      return;
    }

    // Map focus distance to a slider element.
    const input = document.querySelector('input[type="range"]');
    input.min = capabilities.focusDistance.min;
    input.max = capabilities.focusDistance.max;
    input.step = capabilities.focusDistance.step;
    input.value = track.getSettings().focusDistance;

>     input.oninput = function(event) {

Reviewer: ```suggestion
    input.oninput = function (event) {
```

>       track.applyConstraints({

Reviewer: Handle errors
```suggestion
      await track.applyConstraints({
```

>         advanced : [{focusMode: "manual", focusDistance: event.target.value}]

Reviewer: Please don't use `advanced` constraints.
```suggestion
        focusMode: "manual", focusDistance: event.target.value
```

>       });

Reviewer: ```suggestion
        });
      } catch (err) {
        console.error(err);
      }
```

    }
    input.hidden = false;

>   }

Reviewer: ```suggestion
    } catch (err) {
      console.error(err);
    }
  }
```

>   function takePhoto() {
>     imageCapture.takePhoto()
>       .then(blob => {

Reviewer: ```suggestion
  async function takePhoto() {
    try {
      const blob = await imageCapture.takePhoto();
```

        console.log('Photo taken: ' + blob.type + ', ' + blob.size + 'B');

        const image = document.querySelector('img');
        image.src = URL.createObjectURL(blob);

>       })
>       .catch(err => console.error('takePhoto() failed: ', err));

Reviewer: ```suggestion
      } catch (err) {
        console.error(err);
      }
```

  }
```

> ## Capabilities

Reviewer: Should this heading perhaps be "Constrainable Properties"?

> This specification allows web developers to get and set various properties in a modern webcam like `whiteBalanceMode`, `exposureMode`, `focusMode`, `exposureMode`, `pointsOfInterest`, `exposureCompensation`, `exposureTime`, `colorTemperature`, `iso`, `brightness`, `contrast`, `pan`, `saturation`, `sharpness`, `focalDistance`, `tilt`, `zoom` and  `torch` if there is support in the target platform.

Reviewer: Maybe read and constrain instead of get and set?
```suggestion
This specification allows web developers to read and constrain various properties in a modern webcam like `whiteBalanceMode`, `exposureMode`, `focusMode`, `exposureMode`, `pointsOfInterest`, `exposureCompensation`, `exposureTime`, `colorTemperature`, `iso`, `brightness`, `contrast`, `pan`, `saturation`, `sharpness`, `focalDistance`, `tilt`, `zoom` and  `torch` if there is support in the target platform.
```
